packages/caravan-fees/src/btcTransactionComponents.ts

```diff
@@ -0,0 +1,231 @@
+import { Satoshis, BTC, TxOutputType } from "./types";
+import { satoshisToBTC } from "./utils";
+import BigNumber from "bignumber.js";
+
+/**
+ * Abstract base class for Bitcoin transaction components.
+ * Provides common functionality for inputs and outputs.
+ */
+export abstract class BtcTxComponent {
+  protected _amountSats: BigNumber;
+
+  /**
+   * @param amountSats - The amount in satoshis (as a string)
+   */
+  constructor(amountSats: Satoshis) {
+    this._amountSats = new BigNumber(amountSats);
+  }
+
+  /**
+   * Get the amount in satoshis
+   * @returns The amount in satoshis (as a string)
+   */
+  get amountSats(): Satoshis {
+    return this._amountSats.toString();
+  }
+
+  /**
+   * Set  amount in satoshis
+   * @param amountSats - New amount in satoshis (as a string)
+   */
+  set amountSats(value: Satoshis) {
+    this._amountSats = new BigNumber(value);
+  }
+
+  /**
+   * Get the amount in BTC
+   * @returns The amount in BTC (as a string)
+   */
+  get amountBTC(): BTC {
+    return satoshisToBTC(this._amountSats.toString());
+  }
+
+  hasAmount(): boolean {
+    return this._amountSats.isGreaterThan(0);
+  }
+
+  /**
+   * Check if the component is valid
+   * @returns True if valid, false otherwise
+   */
+  abstract isValid(): boolean;
+}
+
+/**
+ * Represents a Bitcoin transaction input
+ */
+export class BtcTxInputTemplate extends BtcTxComponent {
+  private readonly _txid: string;
+  private readonly _vout: number;
+
+  /**
+   * @param params - Input parameters
+   * @param params.txid - Transaction ID of the UTXO
+   * @param params.vout - Output index of the UTXO
+   * @param params.amountSats - Amount in satoshis (as a string)
+   */
+  constructor(params: { txid: string; vout: number; amountSats?: Satoshis }) {
+    super(params.amountSats || "0");
+    this._txid = params.txid;
+    this._vout = params.vout;
+  }
+
+  /** Get the transaction ID */
+  get txid(): string {
+    return this._txid;
+  }
+
+  /** Get the output index */
+  get vout(): number {
+    return this._vout;
+  }
+
+  /**
+   * Check if the input is valid
+   * @returns True if the amount is positive and exists, and txid and vout are valid
+   */
+  isValid(): boolean {
+    return this.hasAmount() && this._txid !== "" && this._vout >= 0;
+  }
+}
+/**
+ * Represents a Bitcoin transaction output
+ */
+export class BtcTxOutputTemplate extends BtcTxComponent {
+  private readonly _address: string;
+  private readonly _type: TxOutputType;
+  private _malleable: boolean = true;
+
+  /**
+   * @param params - Output parameters
+   * @param params.address - Recipient address
+   * @param params.amountSats - Amount in satoshis  (as a string)
+   * @param params.type - Type of output (external or change)
+   */
+  constructor(params: {
```

Comment:
I wonder if external should be locked immediately. External amounts by definition I'd expect to be fixed whereas change outputs by definitions are ones that are malleable until the tx is "done"

Reply:
added check for [that](https://github.com/caravan-bitcoin/caravan/pull/114/commits/6662b2d56bc5f1cef57c403cdc13c2a0ccaac980#diff-db5fc022ab08b92fb2a606342674432f918748c762c9fcb907bc43d3531d512bR173)